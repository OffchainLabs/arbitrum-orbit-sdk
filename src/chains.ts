import { defineChain } from 'viem';
import {
  mainnet,
  holesky,
  arbitrum as arbitrumOne,
  arbitrumNova,
  sepolia,
  arbitrumSepolia,
} from 'viem/chains';

const nitroTestnodeL1 = defineChain({
  id: 1_337,
  network: 'Nitro Testnode L1',
  name: 'nitro-testnode-l1',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8545'],
    },
    public: {
      http: ['http://127.0.0.1:8545'],
    },
  },
  testnet: true,
});

const nitroTestnodeL2 = defineChain({
  id: 412_346,
  network: 'Nitro Testnode L2',
  name: 'nitro-testnode-l2',
  nativeCurrency: { name: 'Ether', symbol: 'ETH', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:8547'],
    },
    public: {
      http: ['http://127.0.0.1:8547'],
    },
  },
  testnet: true,
});

// the L3 deployed as part of the nitro-testnode framework uses a custom fee token
const nitroTestnodeL3 = defineChain({
  id: 333333,
  network: 'Nitro Testnode L3',
  name: 'nitro-testnode-l3',
  nativeCurrency: { name: 'AppTestToken', symbol: 'APP', decimals: 18 },
  rpcUrls: {
    default: {
      http: ['http://127.0.0.1:3347'],
    },
    public: {
      http: ['http://127.0.0.1:3347'],
    },
  },
  testnet: true,
});

export const chains = [
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
] as const;

export {
  // mainnet
  mainnet,
  arbitrumOne,
  arbitrumNova,
  // testnet
  sepolia,
  holesky,
  arbitrumSepolia,
  // local nitro-testnode
  nitroTestnodeL1,
  nitroTestnodeL2,
  nitroTestnodeL3,
};
