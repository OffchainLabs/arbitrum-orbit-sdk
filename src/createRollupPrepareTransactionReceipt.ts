import {
  TransactionReceipt,
  getAbiItem,
  getEventSelector,
  Log,
  decodeEventLog,
  DecodeEventLogReturnType,
} from 'viem';

import { rollupCreatorABI as rollupCreatorV3Dot1ABI } from './contracts/RollupCreator';
import { rollupCreatorABI as rollupCreatorV2Dot1ABI } from './contracts/RollupCreator/v2.1';

import { CoreContracts } from './types/CoreContracts';

function findRollupCreatedEventLog(txReceipt: TransactionReceipt): Log<bigint, number> {
  // v3.1
  const v3Dot1EventSelector = getEventSelector(
    getAbiItem({ abi: rollupCreatorV3Dot1ABI, name: 'RollupCreated' }),
  );
  // v2.1 and v1.1 are the same, so we only need to handle v2.1
  const v2Dot1EventSelector = getEventSelector(
    getAbiItem({ abi: rollupCreatorV2Dot1ABI, name: 'RollupCreated' }),
  );

  const log = txReceipt.logs.find((log) => {
    const topic = log.topics[0];
    return topic === v3Dot1EventSelector || topic === v2Dot1EventSelector;
  });

  if (typeof log === 'undefined') {
    throw new Error(
      `No "RollupCreated" logs found in logs for transaction: ${txReceipt.transactionHash}`,
    );
  }

  return log;
}

type DecodeRollupCreatedEventLogReturnType = DecodeEventLogReturnType<
  // v3.1
  | typeof rollupCreatorV3Dot1ABI
  // v2.1 and v1.1 are the same, so we only need to handle v2.1
  | typeof rollupCreatorV2Dot1ABI,
  'RollupCreated'
>;

function decodeRollupCreatedEventLog(
  log: Log<bigint, number>,
): DecodeRollupCreatedEventLogReturnType {
  let decodedEventLog: DecodeRollupCreatedEventLogReturnType | null = null;

  // try parsing from multiple RollupCreator versions
  [
    // v3.1
    rollupCreatorV3Dot1ABI,
    // v2.1 and v1.1 are the same, so we only need to handle v2.1
    rollupCreatorV2Dot1ABI,
  ].forEach((abi) => {
    try {
      const result = decodeEventLog({ ...log, abi });

      if (result.eventName === 'RollupCreated') {
        decodedEventLog = result;
      }
    } catch (error) {
      // do nothing
    }
  });

  if (decodedEventLog === null) {
    throw new Error(`Failed to decode "RollupCreated" event log: ${log}`);
  }

  return decodedEventLog;
}

export type CreateRollupTransactionReceipt = TransactionReceipt & {
  getCoreContracts(): CoreContracts;
};

export function createRollupPrepareTransactionReceipt(
  txReceipt: TransactionReceipt,
): CreateRollupTransactionReceipt {
  return {
    ...txReceipt,
    getCoreContracts: function () {
      const eventLog = findRollupCreatedEventLog(txReceipt);
      const decodedEventLog = decodeRollupCreatedEventLog(eventLog);

      const { rollupAddress, inboxAddress, ...rest } = decodedEventLog.args;

      return {
        rollup: rollupAddress,
        inbox: inboxAddress,
        ...rest,
        deployedAtBlockNumber: Number(txReceipt.blockNumber),
      };
    },
  };
}
