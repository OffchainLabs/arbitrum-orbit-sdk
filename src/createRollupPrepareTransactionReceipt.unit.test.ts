import { it, expect } from 'vitest';
import { createPublicClient, http } from 'viem';
import { sepolia, arbitrumSepolia } from 'viem/chains';

import { createRollupPrepareTransactionReceipt } from './createRollupPrepareTransactionReceipt';

const client = createPublicClient({
  chain: arbitrumSepolia,
  transport: http(),
});

const sepoliaClient = createPublicClient({
  chain: sepolia,
  transport: http('https://sepolia.gateway.tenderly.co'),
});

// https://sepolia.arbiscan.io/tx/0x5b0b49e0259289fc89949a55a5ad35a8939440a55065d29b14e5e7ef7494efff
it('successfully parses core contracts from a tx receipt on RollupCreator v1.1', async () => {
  const txReceipt = await client.getTransactionReceipt({
    hash: '0x5b0b49e0259289fc89949a55a5ad35a8939440a55065d29b14e5e7ef7494efff',
  });
  expect(createRollupPrepareTransactionReceipt(txReceipt).getCoreContracts()).toMatchSnapshot();
});

// https://sepolia.arbiscan.io/tx/0x77db43157182a69ce0e6d2a0564d2dabb43b306d48ea7b4d877160d6a1c9b66d
it('successfully parses core contracts from a tx receipt on RollupCreator v2.1', async () => {
  const txReceipt = await client.getTransactionReceipt({
    hash: '0x77db43157182a69ce0e6d2a0564d2dabb43b306d48ea7b4d877160d6a1c9b66d',
  });
  expect(createRollupPrepareTransactionReceipt(txReceipt).getCoreContracts()).toMatchSnapshot();
});

// https://sepolia.etherscan.io/tx/0x751f1b2bab2806769f663db2141d434e4d8c9b65bc4a5d7ca10ed597f918191f
it('successfully parses core contracts from a tx receipt on RollupCreator v3.1', async () => {
  const txReceipt = await sepoliaClient.getTransactionReceipt({
    hash: '0x751f1b2bab2806769f663db2141d434e4d8c9b65bc4a5d7ca10ed597f918191f',
  });
  expect(createRollupPrepareTransactionReceipt(txReceipt).getCoreContracts()).toMatchSnapshot();
});
