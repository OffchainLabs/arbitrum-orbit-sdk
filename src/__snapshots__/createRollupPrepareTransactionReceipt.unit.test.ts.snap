// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`successfully parses core contracts from a tx receipt on RollupCreator v1.1 1`] = `
{
  "adminProxy": "0x38E3258488C2341606230E27c4a440e81c987fCF",
  "bridge": "0xac51D5f4B14f520D111D14777e8965b224CE2919",
  "challengeManager": "0x5f4273F26836B3Def34Bc11274Dd7D38200E8E2d",
  "deployedAtBlockNumber": 49500532,
  "inbox": "0x6fDBF2541c1168E6D9f7A3A131204a33Ace085df",
  "nativeToken": "0x0000000000000000000000000000000000000000",
  "outbox": "0x2732C86946c20b9Eba0d7fA0beCcfd7517FCD285",
  "rollup": "0x1644590Fd2223264ea8Cda8927B038CcCFE0Da76",
  "rollupEventInbox": "0x60F938D87f929D1B413BA6bc9d8cA8310710BF1a",
  "sequencerInbox": "0x96bA492C55Af83dfC88D52A1e584e4061716e9e8",
  "upgradeExecutor": "0x5FbdF85F9be59A8084dbD8a7BD457AD3c13C95aE",
  "validatorUtils": "0xB11EB62DD2B352886A4530A9106fE427844D515f",
  "validatorWalletCreator": "0xEb9885B6c0e117D339F47585cC06a2765AaE2E0b",
}
`;

exports[`successfully parses core contracts from a tx receipt on RollupCreator v2.1 1`] = `
{
  "adminProxy": "0x4c6f9229f13cD1922B0E237c5A01CA6466217B11",
  "bridge": "0xFa35d3bEaCE95bfaaDC91EbA202821e952354C54",
  "challengeManager": "0xE1F48151c01B2861365bA22A8F1E676ceBD145Dd",
  "deployedAtBlockNumber": 71259460,
  "inbox": "0x83fac80e23863bAdA976Bddf1aF02B53402e4829",
  "nativeToken": "0x0000000000000000000000000000000000000000",
  "outbox": "0xC062C937CE2f4f7Abb6998bF958B7D67A4348926",
  "rollup": "0x66d0e72952f4f69aF9D33C1B7C31Fa9aCDbCAF63",
  "rollupEventInbox": "0xD75c19D04476dF0B6367b064CBA084EC00038f7d",
  "sequencerInbox": "0x43528b8Be52e8D084F147d167487f361553463b5",
  "upgradeExecutor": "0xcC86ED288eD5DC90Ef9b0e58A12f354840a0B659",
  "validatorUtils": "0x7C100c97a54e2D309a194752Df2f66922A802be3",
  "validatorWalletCreator": "0xFAd2C6Cb969Ab7B18d78BD63e512b650bb70B570",
}
`;

exports[`successfully parses core contracts from a tx receipt on RollupCreator v3.1 1`] = `
{
  "adminProxy": "0xf58519b0A94fd7a12C99c30f6FE984EFea1ac1de",
  "bridge": "0x633dE45a02bB399fA519B8F91f62F9Ef57ACDE52",
  "challengeManager": "0x0938581d28E37EbD23Ef448fd02F8beB206e0383",
  "deployedAtBlockNumber": 8131558,
  "inbox": "0x05F27c4F65abBe0c59a13A924148Da3eafF91C53",
  "nativeToken": "0x0000000000000000000000000000000000000000",
  "outbox": "0x5F54ebfC67fa30565b725e709A5712dcD67C2880",
  "rollup": "0x6677e09F9475Bb66C320Cd50C5db8Ae75D9E42b7",
  "rollupEventInbox": "0x12d28390eB4515e98cd467Cec394BE1a2c79E481",
  "sequencerInbox": "0xdddbBeD40a93985DDC1fbeaF8d588804919eF707",
  "upgradeExecutor": "0x96Eb6Da3FA34494AfDdc3466ffdc9DFc194008c3",
  "validatorWalletCreator": "0x1cB01EA4b15884DEf0e64201C342681B897Db38F",
}
`;
